### DISCUSSION TITLE
Need Help Displaying Many Clipboard Items Quickly in Raycast Extension

### DISCUSSION THREAD

**Discussion by andrewcincotta** [hluk/CopyQ#2543] [Q&A]
I am working on a [CopyQ extension for Raycast](https://github.com/andrewcincotta/copyq-raycast). My extension works as is, but because I am not sure how to utilize CopyQ's terminal command efficiently, I am displaying clipboard items by running "copyq tab {tab} read {index}" where the command is run an index amount of times.

Using the terminal in this fashion takes a while, so I've added default_num_items (set by default to 10). As a result, the extension only displays 10 clipboard items because showing the whole list could take a very long time.

I am looking for someone who can point me in the right direction to list all clipboard items in a directory in a much more efficient way. I can handle things on the Raycast side, but if someone can find a replacement for looping through the "copyq read" command, I can display much more clipboard items while removing the need for a default number of items.

I have looked into [this repo](https://github.com/cjbassi/rofi-copyq) for inspiration, but I am having difficulty comprehending what is being executed.

**Comment by hluk** [comment]
You can print all items in one go with null character (`\0`) as separator:

```bash
copyq tab SOME_TAB 'separator(String.fromCharCode(0)); read.apply(this, [...Array(size()).keys()])'
```

> I have looked into [this repo](https://github.com/cjbassi/rofi-copyq) for inspiration, but I am having difficulty comprehending what is being executed.

They use a script to create an array with all the items and print it as JSON. Here is a simplified version:

```js
var result=[];
for ( var i = 0; i < size(); ++i ) {
    result.push(str(read(i)));
}
JSON.stringify(result);
```

You can pass the script to CopyQ with: `copyq tab SOME_TAB SCRIPT`

Here is a faster version without any loops and empty texts:

```bash
copyq tab SOME_TAB 'JSON.stringify(ItemSelection().select(/./).itemsFormat(mimeText).map(function(x){return str(x)}))'
```

**Reply by andrewcincotta** [reply]
The last command runs very fast, which is great. I am having difficulty storing items in an array properly. Things like new lines and quotes being contained in clipboard items do not transfer over when I try to parse the json and push it to an array. Could you create a javascript function that will properly save each item in a list properly?

**Reply by hluk** [reply]
The `JSON.stringify()` JavaScript function handles encoding the JSON properly (including arrays with multi-line strings containing quotes). You need to parse the output as JSON properly on the other side.

Example with [jq](https://github.com/jqlang/jq) tool:

```bash
❯ copyq -- 'JSON.stringify(["this is \"quoted\"", "this contains\na new line"])'
["this is \"quoted\"","this contains\na new line"]

❯ copyq -- 'JSON.stringify(["this is \"quoted\"", "this contains\na new line"])'| jq -r '.[]'
this is "quoted"
this contains
a new line
```

**Reply by andrewcincotta** [reply]
This works, but can you go back to the original command you made and have it include empty clipboard items. Part of the extension I made selects and pastes the clipboard item, and in order to do that properly, it has to have the correct index from copyq. If there are empty items being omitted, I can't simply use the array's index to determine copyq's actual index

Update:
It may be better to rewrite the command so that it outputs a 2d array/json of the index and content, but whatever you decide is faster or better I will adapt to.

**Reply by hluk** [reply]
With empty items:

```bash
copyq tab SOME_TAB 'JSON.stringify(ItemSelection().selectAll().itemsFormat(mimeText).map(function(x){return x ? str(x) : ""}))'
```

**Reply by andrewcincotta** [reply]
I resolved this by handling all of the parsing in my typescript extension. That required using the original command, which separates items by a null character:
`copyq tab SOME_TAB 'separator(String.fromCharCode(0)); read.apply(this, [...Array(size()).keys()])'`
I found this much easier than worrying about new lines and quotes breaking the parsing of the list/json.

**Comment by andrewcincotta** [comment]
Issue resolved with last reply.
